src/compiler/transformer.ts

```diff
@@ -0,0 +1,334 @@
+                if (node.locals && hasProperty(node.locals, name)) {
+                    // We conservatively include alias symbols to cover cases where they're emitted as locals
+                    if (node.locals[name].flags & (SymbolFlags.Value | SymbolFlags.ExportValue | SymbolFlags.Alias)) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+
+        /**
+         * Generates a name for a node.
+         */
+        function generateNameForNode(node: Node): Identifier {
+            switch (node.kind) {
+                case SyntaxKind.Identifier:
+                    return makeUniqueName((<Identifier>node).text);
+                case SyntaxKind.ModuleDeclaration:
+                case SyntaxKind.EnumDeclaration:
+                    return generateNameForModuleOrEnum(<ModuleDeclaration | EnumDeclaration>node);
+                case SyntaxKind.ImportDeclaration:
+                case SyntaxKind.ExportDeclaration:
+                    return generateNameForImportOrExportDeclaration(<ImportDeclaration | ExportDeclaration>node);
+                case SyntaxKind.FunctionDeclaration:
+                case SyntaxKind.ClassDeclaration:
+                case SyntaxKind.ExportAssignment:
+                    return generateNameForExportDefault();
+                case SyntaxKind.ClassExpression:
+                    return generateNameForClassExpression();
+                default:
+                    return createTempVariable(TempVariableKind.Auto);
+            }
+        }
+
+        function generateNameForModuleOrEnum(node: ModuleDeclaration | EnumDeclaration) {
+            const name = node.name;
+            // Use module/enum name itself if it is unique, otherwise make a unique variation
+            return isUniqueLocalName(name.text, node) ? name : makeUniqueName(name.text);
+        }
+
+        function generateNameForImportOrExportDeclaration(node: ImportDeclaration | ExportDeclaration) {
+            const expr = getExternalModuleName(node);
+            const baseName = expr.kind === SyntaxKind.StringLiteral
+                ? escapeIdentifier(makeIdentifierFromModuleName((<LiteralExpression>expr).text))
+                : "module";
+            return makeUniqueName(baseName);
+        }
+
+        function generateNameForExportDefault() {
+            return makeUniqueName("default");
+        }
+
+        function generateNameForClassExpression() {
+            return makeUniqueName("class");
+        }
+
+        /**
+         * Records a hoisted variable declaration within a lexical environment.
+         */
+        function hoistVariableDeclaration(name: Identifier): void {
+            if (!hoistedVariableDeclarations) {
+                hoistedVariableDeclarations = [];
+            }
+
+            hoistedVariableDeclarations.push(createVariableDeclaration(name));

```

Comment:
I think this is cleaner actually.

In fact, maybe what you'd be better off doing is

```sts
const decl = createVariableDeclaration(name);

if (!hoistedVariableDeclarations) {
     hoistedVariableDeclarations = [decl];
}
else {
    hoistedVariableDeclarations.push(decl);
}